### ISSUE TITLE
Http3Transport draft implementation

### ISSUE THREAD

**Issue by alexhultman** [uNetworking/uWebSockets#1112]
uWS is going to implement the QUIC-based QuicTransport protocol as a complement to WebSocket.

The idea is that since QuicTransport has the concept of URL, it can fit nicely into the app/routes concept that is so central to uWS:

app.qt("/route", {
// whatever behavior here
})

This can be applicable to web messaging that is not in need of reliable & ordered messaging, such as IO-games.

**Comment by adminy** [comment]
any progress update?

**Comment by alexhultman** [comment]
QuicTransport is still highly experimental and not available in any browser by default. This is a long term goal of many months in the future.

**Comment by uasan** [comment]
It is interesting then to compare uWS vs Node.js by QUIC protocol
> The net module provides an implementation of the QUIC protocol. To access it, the Node.js binary must be compiled using the --experimental-quic configuration flag.
https://nodejs.org/dist/latest-v15.x/docs/api/quic.html

**Comment by alexhultman** [comment]
Difference is that uWS will not expose QUIC - it will only expose QuicTransport (the WebTransport server counterpart). So our interfaces will be a lot simpler and high-level for business logic coding.

From gut feeling I will probably use https://github.com/litespeedtech/lsquic (but any QUIC-stack should be usable via configuration).

**Comment by alexhultman** [comment]
It should pretty much look exactly the same as for WebSockets but with a few added features such as sending datagrams that can be lost without causing backpressure - and that SSL is required. Other than this it should look roughly the same (if I'm not entirely wrong here).

**Comment by alexhultman** [comment]
Looks like the IETF just deleted QuicTransport so pretty good I did not start this yet. They are talking about some HTTP3 stuff instead. Let's see where they decide to go with this.

**Comment by leeoniya** [comment]
wasn't it superseded by https://w3c.github.io/webtransport/

also https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Intro_to_RTP

**Comment by alexhultman** [comment]
WebTransport is Http3Transport

https://tools.ietf.org/html/draft-vvv-webtransport-http3-03

**Comment by alexhultman** [comment]
Looks like this library will have to implement Http3 and WebTransport. It kind of maps well to the existing API and has a purpose; Node.js core developers will implement all this really inefficiently (assuming hell hasn't frozen over) so that will be the point of uWS.

**Comment by alexhultman** [comment]
https://caniuse.com/?search=http3

Http3 is still heavily unsupported. This really won't become a thing for many years to come. And then they also need to finish/add this WebTransport thing. And then developers need to actually adopt it!

So v19 will be really just a polish of what people need **today**. Maybe it is time to add QUIC to uSockets soon, but more than that is rushing too fast.

**Comment by zorro-fr24** [comment]
> Http3 is still heavily unsupported. This really won't become a thing for many years to come.

I think it's not so far off ( provided they can lock the standard down ). All the main browser vendors have working draft implementations that can be enabled by feature flags, and major network providers like cloudflare already support it on their edge:
![image](https://user-images.githubusercontent.com/7267001/108314082-9e150c00-71b9-11eb-860f-11201b947cc6.png)

**Comment by alexhultman** [comment]
I'm looking at QUIC in uSockets first of all. Once that is done, the rest can be built easily. And even in case of emergency you can really just take the HTTP/3 support of the QUIC-stack itself and just expose. But I probably want to write my own layer on QUIC. It depends how the benchmarks are. It could simply be the case that these stacks are well-written.

**Comment by uasan** [comment]
Probably then it will be necessary to rename the library to `uWebTransport`, because the websockets will not be relevant )

**Comment by lucsoft** [comment]
i don't thing websocket will be going away soon or later

but im still hyped for WebTransport

**Comment by alexhultman** [comment]
> > Http3 is still heavily unsupported. This really won't become a thing for many years to come.
> 
> I think it's not so far off ( provided they can lock the standard down ). All the main browser vendors have working draft implementations that can be enabled by feature flags, and major network providers like cloudflare already support it on their edge:
> ![image](https://user-images.githubusercontent.com/7267001/108314082-9e150c00-71b9-11eb-860f-11201b947cc6.png)

Http3 itself is still a draft as of today. Http3Transport - the WebTransport-on-Http3 is like draft squared - they are still arguing about the standard. So this will all take many years. Http2 is kind of mainstream today, some 5 years after introduction. But still many services run on regular HTTP1 internally.

**Comment by uasan** [comment]
It looks like they agreed on the API.
First Public Working Draft of WebTransport.
https://www.w3.org/blog/news/archives/9043

**Comment by uasan** [comment]
Stateless by design, it's degradation relative to WebSocket (
> For instance, none of the transports defined in this document automatically send cookies, support HTTP authentication or caching invalidation mechanisms. Since they use TLS, they may support TLS session tickets

https://www.w3.org/TR/2021/WD-webtransport-20210504/#state-persistence
